rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -586,7 +586,10 @@ RecorderImpl::create_subscription(
   }
 #endif
 
-  if (record_options_.use_sim_time) {
+  static bool rwm_has_received_timestamp_support =
+    std::string(rmw_get_implementation_identifier()).find("rmw_cyclonedds") == std::string::npos;
```

Comment:
I have two thoughts about this.

1.  Does rmw_connextdds properly set the receive_timestamp?
2.  This fix seems awfully specific to `rmw_cyclonedds`.  What about other RMWs that don't happen to set the receive_timestamp properly?  Would it be better to instead detect whether `receive_timestamp` is zero, and then do a fallback behavior if it is?

Reply:
@clalancette 
- As regards 1. It seems yes since CI is green, and I've added test coverage.
- As regards 2. The changes in this PR is a workaround. The proper fix would be to set received_timestamp at least on RMW layer i.e.,
https://github.com/ros2/rmw_cyclonedds/blob/c6dbe24b2f2be87cf8e4750d89501657ab83566f/rmw_cyclonedds_cpp/src/rmw_node.cpp#L3280-L3282
```
  message_info->source_timestamp = info.source_timestamp;
  // TODO(iluetkeb) add received timestamp, when implemented by Cyclone
  message_info->received_timestamp = 0;
```
- As regards:
> What about other RMWs that don't happen to set the receive_timestamp properly? Would it be better to instead detect whether receive_timestamp is zero, and then do a fallback behavior if it is?

Well we have API where RMW or underlying transport implementation shall settle up value for received timestamp in message info. If they do not do this - this is should considered as bug on their side. I would understand if `send_timestamp` could be not set because it might not be available but with receive - no problem to set it.
- As regards the idea of checking for a timestamp for each message, and if it is zero, overwriting it with our timestamp - IMO, it looks like a hack, and we shouldn't do this.

Reply:
I only was thinking of adding a warning message on the subscription that RMW doesn't support `received_timestamp`

Reply:
@clalancette `rmw_connextdds` properly set received and send timestamps among a case with Windows which we have already workaounded in the same function a few lines above (570-587)
https://github.com/ros2/rmw_connextdds/blob/d2bd90826e19097a429ba150b12e67d3822c1ccc/rmw_connextdds_common/src/common/rmw_impl.cpp#L1970-L1983
```cpp
rmw_connextdds_message_info_from_dds(
  rmw_message_info_t * const to,
  const DDS_SampleInfo * const from)
{
  rmw_connextdds_ih_to_gid(from->publication_handle, to->publisher_gid);
// Message timestamps are disabled on Windows because RTI Connext DDS
// does not support a high enough clock resolution by default (see: _ftime()).
#if !RTI_WIN32
  to->source_timestamp = dds_time_to_u64(&from->source_timestamp);
  to->received_timestamp = dds_time_to_u64(&from->reception_timestamp);
#else
  to->source_timestamp = 0;
  to->received_timestamp = 0;
#endif /* !RTI_WIN32 */
```

Reply:
Doesn't this workaround on Connext Windows mean that rosbag2 won't be able to playback data on that combination?  In other words, should we do a fix to `rmw_connextdds` as well with a similar fix to https://github.com/ros2/rmw_cyclonedds/pull/491 ?

Reply:
As regards:
> Doesn't this workaround on Connext Windows mean that rosbag2 won't be able to playback data on that combination?

No. the workaround is correct and rosbag2 will playback with `connext_dds` it will take received_timestamp from node clock on Windows and in all other cases it will be settled up by `rmw_connextdds` directly.

From one side, yes we can implement the same fix as we currently do for `rmw_cyclonedds`. However, from another side, it is not an issue currently and a workaround is already in place. I wouldn't bother for it now. At least I have more important things to do.

Reply:
All right, that's good enough for now.  Let's just concentrate on fixing rmw_cyclonedds then.